Plugins/GeoModel/src/GeoModelModuleSplitter.cpp

```diff
@@ -0,0 +1,86 @@
+//
+// Copyright (C) 2024 CERN for the benefit of the Acts project
+//
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at http://mozilla.org/MPL/2.0/.
+
+#include "Acts/Plugins/GeoModel/GeoModelModuleSplitter.hpp"
+
+#include "Acts/Surfaces/AnnulusBounds.hpp"
+#include "Acts/Surfaces/DiscSurface.hpp"
+#include "Acts/Utilities/Logger.hpp"
+
+#include <iostream>
+
+namespace Acts {
+
+auto GeoModelModuleSplitter::split(DetectorElementPtr detElement,
+                                   const Acts::GeometryContext &gctx) const
+    -> std::optional<std::vector<DetectorElementPtr>> {
+  const auto &logger = *m_logger;
+
+  auto surface = detElement->surface().getSharedPtr();
+  auto annulusBounds = dynamic_cast<const AnnulusBounds *>(&surface->bounds());
+
+  if (annulusBounds == nullptr) {
+    ACTS_DEBUG("Not annulus bounds");
+    return std::nullopt;
+  }
+
+  std::optional<std::vector<DetectorElementPtr>> result;
+
+  for (auto [patternName, radii] : m_splitPatterns) {
+    if ((std::abs(radii.front() - annulusBounds->rMin()) > m_tolerance) ||
+        (std::abs(radii.back() - annulusBounds->rMax()) > m_tolerance)) {
+      ACTS_VERBOSE("Skip pattern '" << patternName << "' for element '"
+                                    << detElement->databaseEntryName());
+      continue;
+    }
+
+    radii.front() = annulusBounds->rMin();
+    radii.back() = annulusBounds->rMax();
+
+    ACTS_DEBUG("Accept pattern '" << patternName << "' for element '"
+                                  << detElement->databaseEntryName());
+
+    result.emplace();
+    result->reserve(radii.size() - 1);
+    const auto origValues = annulusBounds->values();
+
+    for (auto i = 0ul; i < radii.size() - 1; ++i) {
+      std::array<double, AnnulusBounds::eSize> values;
+      std::copy(origValues.begin(), origValues.end(), values.begin());
+      values[AnnulusBounds::eMinR] = radii[i];
+      values[AnnulusBounds::eMaxR] = radii[i + 1];
+
+      ACTS_VERBOSE("Make new annulus bounds: " << [&]() {
+        std::stringstream ss;
+        for (auto v : values) {
+          ss << v << " ";
+        }
+        return ss.str();
+      }());
+      auto bounds = std::make_shared<AnnulusBounds>(values);
+
+      auto newDetElement =
+          GeoModelDetectorElement::createDetectorElement<DiscSurface>(
+              detElement->physicalVolume(), bounds, detElement->transform(gctx),
+              detElement->thickness());
+      newDetElement->setDatabaseEntryName(detElement->databaseEntryName());
+
+      result->push_back(newDetElement);
+    }
+
+    return result;
+  }
+
+  ACTS_WARNING("Could not split '" << detElement->databaseEntryName()
+                                   << "' (rmin: " << annulusBounds->rMin()
+                                   << ", rmax: " << annulusBounds->rMax()
+                                   << ")");
+  return result;
```

Comment:
I think this is the `null` case. Does this make sense as a warning? What are the valid cases? If it's valid, it shouldn't be a WARNING, if it's invalid, it should be an error.

Could the return value just contain the original detector element in case it wasn't split, if that's not an error condition?

Reply:
Absolutely, if it's an error, you can probably throw an exception and not return anything at all, right?